src/java.base/unix/native/libnio/fs/UnixCopyFile.c

```diff
@@ -50,6 +51,19 @@ static void throwUnixException(JNIEnv* env, int errnum) {
     }
 }
 
+#if defined(_ALLBSD_SOURCE)
+int fcopyfile_callback(int what, int stage, copyfile_state_t state,
+    const char* src, const char* dst, void* cancel)
+{
+    if (what == COPYFILE_COPY_DATA && stage == COPYFILE_PROGRESS &&
+        *((int*)cancel) != 0) {
+        // errno will be set to ECANCELED
+        return COPYFILE_QUIT;
+    }
+    return COPYFILE_CONTINUE;
+}
+#endif
```

Comment:
I wasn't aware of this fcopyfile but it seems to be a good match. At some point I think we will have to split up UnixCopyFile to make it easier to maintain the different implementations. That might be the opportunity to do more than COPYFILE_DATA, meaning we could get it to copy the meta data too.

I think we should try to re-format the callback a bit to make it easier to distinguish the conditions in the if-statement from the body.  I probably should re-format the declaration too to get it a more more consistent/readable.

Reply:
I agree that it would be good to split it up but as part of a yet to be filed issue.

Would you please elaborate on reformatting the callback? Thanks.

Reply:
I should have been clearer. L58 splits the conditions on two lines with the second line indented 4 spaces so it looks like the 3rd condition is in the block. I think you should re-format it to make it easier to read.

Reply:
This could be made into one line, no split, or three lines. Otherwise an additional `if` would be needed.